Source/Core/Core/IOS/Network/KD/NWC24DL.cpp

```diff
@@ -0,0 +1,150 @@
+// Copyright 2022 Dolphin Emulator Project
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include "Core/IOS/Network/KD/NWC24DL.h"
+
+#include "Common/CommonPaths.h"
+#include "Common/CommonTypes.h"
+#include "Common/Logging/Log.h"
+#include "Common/Swap.h"
+#include "Core/IOS/FS/FileSystem.h"
+#include "Core/IOS/Uids.h"
+
+namespace IOS::HLE::NWC24
+{
+constexpr const char DL_LIST_PATH[] = "/" WII_WC24CONF_DIR "/nwc24dl.bin";
+
+NWC24Dl::NWC24Dl(std::shared_ptr<FS::FileSystem> fs) : m_fs{std::move(fs)}
+{
+  ReadDlList();
+}
+
+void NWC24Dl::ReadDlList()
+{
+  if (const auto file = m_fs->OpenFile(PID_KD, PID_KD, DL_LIST_PATH, FS::Mode::Read))
+  {
+    if (file->Read(&m_data, 1))
+    {
+      const s32 file_error = CheckNwc24DlList();
+      if (file_error)
+        ERROR_LOG_FMT(IOS_WC24, "There is an error in the DL list for WC24: {}", file_error);
+
+      return;
+    }
+  }
+}
+
+s32 NWC24Dl::CheckNwc24DlList() const
+{
+  // 'WcDl' magic
+  if (Magic() != 0x5763446c)
+  {
+    ERROR_LOG_FMT(IOS_WC24, "DL list magic mismatch");
+    return -1;
+  }
+
+  if (Version() != 1)
+  {
+    ERROR_LOG_FMT(IOS_WC24, "DL list version mismatch");
+    return -1;
+  }
+
+  return 0;
+}
+
+void NWC24Dl::WriteDlList() const
+{
+  constexpr FS::Modes public_modes{FS::Mode::ReadWrite, FS::Mode::ReadWrite, FS::Mode::ReadWrite};
+  m_fs->CreateFullPath(PID_KD, PID_KD, DL_LIST_PATH, 0, public_modes);
+  const auto file = m_fs->CreateAndOpenFile(PID_KD, PID_KD, DL_LIST_PATH, public_modes);
+
+  if (!file || !file->Write(&m_data, 1))
+    ERROR_LOG_FMT(IOS_WC24, "Failed to open or write WC24 DL list file");
+}
+
+void NWC24Dl::ResetConfig()
+{
+  // m_fs->Delete(PID_KD, PID_KD, CONFIG_PATH);
+  // TODO: implement
+}
+
+std::string NWC24Dl::GetDownloadURL(u16 entryIndex)
```

Comment:
While it's unlikely to occur, all these functions should likely check that the entry index is something valid.

Example:

```
if (entryIndex < NWC24Dl::MAX_ENTRIES)
{
  ERROR_LOG_FMT(IOS_WC24, "Failed to get download url, entry index out of bounds", entryIndex);
  return;
}
```

Reply:
This could be a good safety thing to have, but I do not know of any case where the entry index comes from the network, from what I know it is set by the channel.

Reply:
Ah sorry, read too quickly, assumed `HandleNWC24DownloadNowEx` 's request came from a network packet.  Not as important then.